package.json

```diff
@@ -2,14 +2,14 @@
   "name": "nbgitpuller",
   "version": "0.10.1",
   "description": "`nbgitpuller`",
-  "devDependencies": {
+  "dependencies": {
+    "css-loader": "^6.2.0",
     "jquery": "^3.6.0",
     "webpack": "^5.45.1",
     "webpack-cli": "^4.7.2",
     "xterm": "^4.13.0",
     "xterm-addon-fit": "^0.5.0",
-    "css-loader": "^6.2.0",
-    "style-loader": "^3.2.1"
+    "mini-css-extract-plugin": "^2.2.0"
   },
```

Comment:
The actual dependencies of the source code part of nbgitpuller is can be structured like this.

```json
  "dependencies": {
    "jquery": "^3.6.0",
    "xterm": "^4.13.0",
    "xterm-addon-fit": "^0.5.0"
  },
  "devDependencies": {
    "css-loader": "^6.2.0",
    "css-minimizer-webpack-plugin": "^3.0.2",
    "mini-css-extract-plugin": "^2.2.0",
    "webpack": "^5.45.1",
    "webpack-cli": "^4.7.2"
  },
```

Currently in the main branch, but not yet released, jquery, xterm, and xterm-addon-fit are listed under `devDependencies`. That makes me think that it has been fine to rely on these libraries being available from other locations for anyone having tested nbgitpuller in action so far, because anything in `devDependencies` doesn't enter the .js file end users installing nbgitpuller get.

```json
  "devDependencies": {
    "jquery": "^3.6.0",
    "webpack": "^5.45.1",
    "webpack-cli": "^4.7.2",
    "xterm": "^4.13.0",
    "xterm-addon-fit": "^0.5.0",
    "css-loader": "^6.2.0",
    "style-loader": "^3.2.1"
  },
```

I tried doing some research:
- Should nbgitpuller assume the website where it runs have loaded the required javascript libraries already, making it okay to not ship it in its own .js bundle?
- What other .js bundles would have jquery, xterm, xterm-addon-fit that we could rely on indirectly?
  jupyter_server: [package.json](https://github.com/jupyter-server/jupyter_server/blob/master/package.json)
  notebook: [package.json](https://github.com/jupyter/notebook/blob/master/package.json), [bower.json](https://github.com/jupyter/notebook/blob/HEAD/bower.json).
  jupyterlab: [package.json](https://github.com/jupyterlab/jupyterlab/blob/master/package.json#L105)
  jupyterlab - terminal: [package.json](https://github.com/jupyterlab/jupyterlab/blob/HEAD/packages/terminal/package.json)
  jupyterlab: [jquery doesn't seem to ship with it](https://github.com/jupyterlab/jupyterlab/search?q=jquery)

@bollwyvl do you have ideas on how / what to package given that nbgitpuller is to support being run on a jupyter_server exposed notebook UI and jupyterlab UI as well as a notebook server exposing a notebook UI or jupyterlab UI?

Note if my question doesn't make sense, it is because I don't understand this well enough to state a question yet :grin:

Reply:
The `(dev)Dependencies` distinction is kinda meaningless in this setting, as they both get smooshed together in `node_modules` and webpack puts whatever it wants into your resultant bundle unless you tell it otherwise.

While `notebook`'s assets are available (the clock is ticking), in that UI (which i _assume_ is just the interstitial progress bar), you'd probably want to bundle just the nbgitpuller UI, if anything, and use `requirejs` to get at your jquery, xterm, etc.

As a `jupyterlab` extension, you'd likely want to fire up one of the [cookiecutters](https://github.com/jupyterlab/extension-cookiecutter-ts), which replace the requirejs/UMD module approach with a webpack-specific one that is better able to handle versions,  singletons, duplicate modules, etc. If you don't need jquery, don't bring it! Also, on closer inspection, I see that the `xterm` depdendency is _not_ hoisted to `sharedPackages`, so in that environment you'd probably want to use the renderers, etc. from the `@jupyterlab/terminal`, rather than bringing in _another_ copy of xterm.

If there was significant overlap between these two implementations, the project could actually be divided into a set of npm packages (multiple `package.json`s), managed with `lerna`,  where:
- the "core" ui would be consumed by
  - the notebook app
  - the lab extension

These could even be _published_ back to npm, which could potentially leave some extension points open (at least in lab) that could be extended by third party extensions.

Reply:
@bollwyvl this is neither a notebook extension nor a lab extension! It is a server extension that displays a page by itself. It used to depend on fetching assets from classic notebook along with the templates defined by classic notebook, but that meant it didn't work with jupyter_server (https://github.com/jupyterhub/nbgitpuller/issues/176). So we've removed that and just use webpack. So it should now hopefully have no interaction with lab or notebook clientside bits. I mostly just don't want us to ship half a meg of JS on each page load...

Reply:
> The (dev)Dependencies distinction is kinda meaningless in this setting, as they both get smooshed together in node_modules and webpack puts whatever it wants into your resultant bundle unless you tell it otherwise.

Aaaah so due to these combined, we end up having all our required dependencies built into bundle.js created by webpack, while the rest are just node_modules stuff that will be ignored.

https://github.com/jupyterhub/nbgitpuller/blob/58e61f6f74e6ce034e48b20738012fcf034087f8/webpack.config.js#L20-L25

https://github.com/jupyterhub/nbgitpuller/blob/58e61f6f74e6ce034e48b20738012fcf034087f8/nbgitpuller/static/js/index.js#L1-L2

Reply:
Yep. So if you want shiny terminal, you must pay the 200kb. jquery, on the other hand, could be dropped, and contributes _just_ as much. Unless you're supporting IE11, you probably don't need it anymore, and since you're not relying on _someone else's jquery_, there's nobody else to blame for the 200kb. 

![Screenshot from 2021-08-28 12-06-53](https://user-images.githubusercontent.com/45380/131225445-377380b0-d96b-408c-a50b-788a17d6f7d7.png)


But either way: i'd save some pain, and just `import $ from 'jquery'` rather than relying on magic webpack stuff.

A way to get _perceptually_ improved performance would be to defer loading `xterm` until it's actually _asked_ for, e.g.

```js
function userActuallyAsksForLogs() {
  const {Terminal} = await import('xterm');
  // do something with the logs
}
```

If jquery were dropped, and xterm only loaded/drawn if needed, the size of the initial load would be _substantially_ improved.

Reply:
If this means we can't drop any size at all without rewriting to remove our jquery dependency, then I don't think this is a blocker for 1.0 of nbgitpuller.

Reply:
Is it worth reviewing what xtermjs is used for in this extension, and whether it's needed? One of the main features of xtermjs is supporting interactive input, is this something nbgitpuller requires now (or in the future)?

Reply:
Kinda seems like a 1.0 is a really _good_ time to address legacy dependencies and performance...